apps/opencs/view/render/terrainshapemode.cpp

```diff
@@ -0,0 +1,1217 @@
+                        else
+                            v2[2] = 0;
+                    }
+
+                    normal[1] = v1[2]*v2[0] - v1[0]*v2[2];
+                    normal[0] = v1[1]*v2[2] - v1[2]*v2[1];
+                    normal[2] = v1[0]*v2[1] - v1[1]*v2[0];
+
+                    hyp = sqrt(normal[0]*normal[0] + normal[1]*normal[1] + normal[2]*normal[2]) / 127.0f;
+
+                    normal[0] /= hyp;
+                    normal[1] /= hyp;
+                    normal[2] /= hyp;
+
+                    landNormalsNew[(j * ESM::Land::LAND_SIZE + i) * 3 + 0] = normal[0];
+                    landNormalsNew[(j * ESM::Land::LAND_SIZE + i) * 3 + 1] = normal[1];
+                    landNormalsNew[(j * ESM::Land::LAND_SIZE + i) * 3 + 2] = normal[2];
+                }
+            }
+            if (allowLandShapeEditing(cellId) == true) pushNormalsEditToCommand(landNormalsNew, document, landTable, cellId);
+        }
+        undoStack.endMacro();
+        mAlteredCells.clear();
+
+        if (CSVRender::PagedWorldspaceWidget *paged = dynamic_cast<CSVRender::PagedWorldspaceWidget *> (&getWorldspaceWidget()))
+            paged->resetAllAlteredHeights();
+    }
+}
+
+
+void CSVRender::TerrainShapeMode::dragAborted()
+{
+    if (CSVRender::PagedWorldspaceWidget *paged =
+        dynamic_cast<CSVRender::PagedWorldspaceWidget *> (&getWorldspaceWidget()))
+    {
+        paged->resetAllAlteredHeights();
+        mTotalDiffY = 0;
+    }
+}
+
+void CSVRender::TerrainShapeMode::dragWheel (int diff, double speedFactor)
+{
+}
+
+void CSVRender::TerrainShapeMode::editTerrainShapeGrid(const std::pair<int, int>& vertexCoords, bool dragOperation)
+{
+    int r = mBrushSize / 2;
+    if (r == 0) r = 1; // Prevent division by zero later, which might happen when mBrushSize == 1
+
+    std::string cellId = CSMWorld::CellCoordinates::vertexGlobalToCellId(vertexCoords);
+    CSMWorld::CellCoordinates cellCoords = CSMWorld::CellCoordinates::fromId(cellId).first;
+
+    if (CSVRender::PagedWorldspaceWidget *paged =
+        dynamic_cast<CSVRender::PagedWorldspaceWidget *> (&getWorldspaceWidget()))
+    {
+        if (mShapeEditTool == 0) paged->resetAllAlteredHeights();
+    }
+
+    if(allowLandShapeEditing(cellId)==true)
+    {
+        if (mBrushShape == 0)
+        {
+            int x = CSMWorld::CellCoordinates::vertexGlobalToInCellCoords(vertexCoords.first);
+            int y = CSMWorld::CellCoordinates::vertexGlobalToInCellCoords(vertexCoords.second);
+            if (mShapeEditTool == 0) alterHeight(cellCoords, x, y, mTotalDiffY);
+            if (mShapeEditTool == 1 || mShapeEditTool == 2) alterHeight(cellCoords, x, y, mShapeEditToolStrength);
+            if (mShapeEditTool == 3) smoothHeight(cellCoords, x, y, mShapeEditToolStrength);
+            if (mShapeEditTool == 4) flattenHeight(cellCoords, x, y, mShapeEditToolStrength, mTargetHeight);
```

Comment:
Consider to use `enum` for `mShapeEditTool` type with some meaningful element names.

Reply:
Committed a bugged version of this, but I fixed the bug in my local branch. I'll commit along other changes later to prevent extra CI builds.